### PR TITLE
Refactor thresholding/dithering API's + bug fixes

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Prompted by the following question on StackOverflow regarding the setting of custom colors during binarization I set about improving our binary threshold methods. This grew arms and legs as I realized the current API's for thresholding didn't cut it and that our ordered dither matrices were incorrect.

https://stackoverflow.com/questions/48265710/transparent-background-using-imagesharps-binarythreshold

This PR:

- Updates `BinaryThreshold` method and other binarization methods to allow the setting of custom colors.
- Renames existing dithering methods that employ binarization to use a `Binary` prefix. e.g `BinaryDiffuse`, `BinaryDither`
- Adds new `Dither` and `Diffuse` methods that accept a custom palette.
- Added `StevensonArceDiffuser`
- Made static version of `KnownDiffusers` to house all known diffusers
- Made static version of `KnownDitherers` to house all known dither matrices
- Added testable `OrderedDitherFactory` that can create thresholding matrices of arbitrary side length
- Added various known ordered dither matrices based on the factory code.
- Reduces allocations and simplifies the static WebSafe color palette.
- Other bug fixes and API improvements.

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
We can now do cool things like pass a custom palette to an ordered dither matrix

Bayer8x8 dither pith web safe palette

Or pass custom colors to all our thresholders

Bayer4x4 binary dither with custopm colors

**Comment by antonfirsov** [comment]
These images remind me of the good-old CGA days! ^^
image

Hope I can do a proper review later this week.

**Comment by antonfirsov** [comment]
@JimBobSquarePants When I try to run multiple `DitherTests` in parallel (only DitherTests!), they all fail with `NullReferenceException`. My stack trace is ending with:

```
   at SixLabors.ImageSharp.PixelFormats.NamedColors`1.GetWebSafePalette() in c:\dev\ImageSharp\_common\src\ImageSharp\PixelFormats\NamedColors{TPixel}.cs:line 733
   at SixLabors.ImageSharp.PixelFormats.NamedColors`1..cctor() in c:\dev\ImageSharp\_common\src\ImageSharp\PixelFormats\NamedColors{TPixel}.cs:line 728
```
We should back the `WebSafePalette` property with `Lazy<T>`.

**Comment by antonfirsov** [comment]
Aaaaand fixed it!

I unfamiliar with the math behind this stuff, but those images look cool, so it must be correct! :)

I think it would be good to have some regression testing by calling `.CompareToReferenceImage()` for a well selected subset of input images (to make future optimizations safe!). I really don't know which one should we choose however.

> Or pass custom colors to all our thresholders

Would be nice to have a test case demonstrating it. (Or at least I wasn't able to find the pink bike in the test output.)

File: src/ImageSharp/Processing/Processors/Dithering/PaletteDitherProcessorBase.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            float leastDistance = int.MaxValue;
+            float secondLeastDistance = int.MaxValue;
+            var vector = pixel.ToVector4();
+
+            var closest = default(TPixel);
+            var secondClosest = default(TPixel);
+            for (int index = 0; index < colorPalette.Length; index++)
+            {
+                TPixel temp = colorPalette[index];
+                float distance = Vector4.Distance(vector, temp.ToVector4());
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/Processors/Dithering/PaletteDitherProcessorBase.cs]: If you are looking for a minimum (and not actually a distance value), it's enough to call Vector4.DistanceSquared() to spare an expensive square root calculation.

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Dithering/PaletteDitherProcessorBase.cs]: We're looking for the distance here, basically we want to find the two closest colors to the given one.

**Comment by JimBobSquarePants** [comment]
I do plan on adding reference images, I just wanted to wait after we've cleaned up all our PR's a little.

**Comment by JimBobSquarePants** [comment]
I'm gonna merge this now. I'm happy with it and don't want to block future work.

**Pull request merged**